spring-batch-samples/src/test/java/org/springframework/batch/sample/RemoteChunkingJobFunctionalTests.java

```diff
@@ -0,0 +1,172 @@
+
+import org.junit.Assert;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.amqp.core.AmqpTemplate;
+import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.batch.core.ExitStatus;
+import org.springframework.batch.core.Job;
+import org.springframework.batch.core.JobExecution;
+import org.springframework.batch.core.configuration.annotation.JobBuilderFactory;
+import org.springframework.batch.core.step.tasklet.TaskletStep;
+import org.springframework.batch.integration.chunk.RemoteChunkingMasterStepBuilderFactory;
+import org.springframework.batch.integration.config.annotation.EnableBatchIntegration;
+import org.springframework.batch.item.ItemReader;
+import org.springframework.batch.item.support.ListItemReader;
+import org.springframework.batch.sample.config.JobRunnerConfiguration;
+import org.springframework.batch.sample.remotechunking.WorkerConfiguration;
+import org.springframework.batch.test.JobLauncherTestUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.integration.amqp.dsl.Amqp;
+import org.springframework.integration.channel.DirectChannel;
+import org.springframework.integration.channel.QueueChannel;
+import org.springframework.integration.config.EnableIntegration;
+import org.springframework.integration.dsl.IntegrationFlow;
+import org.springframework.integration.dsl.IntegrationFlows;
+import org.springframework.messaging.PollableChannel;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+import java.util.Arrays;
+
+/**
+ * <p>Ensure a RabbitMQ instance is running, modifying default.amqp.properties if needed.
+ * Execute the {@link WorkerConfiguration#main(java.lang.String[])} method to start a worker.</p>
+ *
+ * <p>The test in this class is ignored just because there must be a worker JVM
+ * running for the test to work. Un-comment the "@Ignore" annotation and run the test
+ * after starting at least one worker.</p>
+ *
+ * <p>The master step of the tested job will read data and send chunks to the worker(s)
+ * for processing and writing.</p>
+ */
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes = {
+		JobRunnerConfiguration.class,
+		RemoteChunkingJobFunctionalTests.MasterConfiguration.class})
+public class RemoteChunkingJobFunctionalTests {
+
+	@Autowired
+	private JobLauncherTestUtils jobLauncherTestUtils;
+
+	@Ignore("This test is ignored because a worker JVM should be running to process chunks")
```

Comment:
Why can't we launch the other JVM?

Reply:
We can launch the other JVM but the issue is when to stop it? Since on the worker side we don't have a job/step, I could not use a listener for example to stop the JVM. It's technically possible to stop the JVM once all chunks received from the master are processed, but this technical code will "pollute" the example and may disturb the user from the main purpose of the sample.

I wrote a unit test on the master side to be consistent with other samples. But AFAIK, this is the first sample requiring two JVMs (the amqp sample does not wait for the other jvm, it just sends some messages to a queue and dies). Since it's a sample, it's probably better to write a main method on the master side too and remove this unit test. What do you think?

Reply:
I'd use `@Before` and `@After` to manage the worker.  That keeps the management of the other JVMs separate from the test code.